docs/jep/26069-pytree-dict-key-order.md

```diff
@@ -0,0 +1,131 @@
+(jep-pytree-key-order)=
+# JEP 26069: PyTree preservation of dict key order
+
+A frequent JAX feature request/bug report (e.g. {jax-issue}`#4085`, {jax-issue}`#5948`,
+{jax-issue}`#8419`, {jax-issue}`#11817`, {jax-issue}`#24398`, {jax-issue}`#)
+concerns the fact that JAX {ref}`PyTrees <pytrees>` do not preserve dictionary key
+order; for example:
+```python
+>>> dct = {'b': 1, 'c': 2, 'a': 3}
+>>> print(list(dct.keys()))
+['b', 'c', 'a']
+
+>>> import jax
+>>> leaves, tree = jax.tree.flatten(dct)
+>>> dct2 = jax.tree.unflatten(tree, leaves)
+>>> print(list(dct2.keys()))
+['a', 'b', 'c']
+```
+The keys of the unflattened dict are sorted, and this fact sometimes
+surprises users.
+
+This JEP explores options for modifying {mod}`jax.tree_util` in order
+to preserve dict key order.
+
+## Why sort keys at all?
+
+The root of this behavior is the fact that `tree_flatten` sorts dict keys:
+```python
+>>> import jax
+>>> dct = {'b': 'b_leaf', 'c': 'c_leaf', 'a': 'a_leaf'}
+>>> leaves, treedef = jax.tree.flatten(dct)
+>>> print(leaves)
+['a_leaf', 'b_leaf', 'c_leaf']
+```
+Prior to Python 3.7, Python dicts did not guarantee preservation of insertion
+order in standard dicts, and so when JAX was first implemented this sorting was
+necessary to ensure predictable and repeatable runtime behavior.
+
+In the era of order-preserving dicts, key sorting still provides a benefit in
+JAX; namely it means the JIT cache is insensitive to the insertion order of
+its inputs.
+```python
+>>> @jax.jit
+... def f(x):
+...   print('tracing')
+...   return x['a'] + x['b']
+...
+>>> dct1 = {'a': 1, 'b': 2}
+>>> print("result:", f(dct1))  # initial trace
+tracing
+result: 3
+
+>>> dct2 = {'b': 1, 'a': 2}
+>>> print("result:", f(dct1))  # hits cache: no trace
+result: 3
+```
+Reordering dict keys doesn't lead to a cache miss, and this is a very good
+thing, becuase it avoids unexpected recompilations.
+
+## Possible approaches for preserving key order
+
+### 1. Preserve order when flattening
```

Comment:
If possible, you should add an example explaining!

Reply:
Is the example in the first sentence not clear enough?

Reply:
Ohh sorry, I missed that.